src/Domain/CoordinateMaps/BulgedCube.cpp

```diff
@@ -0,0 +1,270 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Domain/CoordinateMaps/BulgedCube.hpp"
+
+#include "DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+namespace CoordinateMaps {
+BulgedCube::BulgedCube(const double radius, const double sphericity,
+                       const bool with_equiangular_map)
+    : radius_(radius),
+      sphericity_(sphericity),
+      with_equiangular_map_(with_equiangular_map) {
+  ASSERT(radius > 0, "The radius of the cube must be greater than zero");
+}
+
+template <typename T>
+std::array<std::decay_t<tt::remove_reference_wrapper_t<T>>, 3> BulgedCube::
+operator()(const std::array<T, 3>& x) const noexcept {
+  using ReturnTypeElement = std::decay_t<tt::remove_reference_wrapper_t<T>>;
+  const ReturnTypeElement& xi = x[0];
+  const ReturnTypeElement& eta = x[1];
+  const ReturnTypeElement& zeta = x[2];
+  const ReturnTypeElement cap_xi =
```

Comment:
You are copying half the time, use ` const&` Do this on your other PRs too.

Reply:
The last time this came up (in Wedge3D) we determined that that change didn't help and the copy happened anyway.

Reply:
Was I one of the "we"? I remember discussing this, but don't remember this conclusion. Is it because the common type imposed by the ternary is not a reference that this occurs? If so would it be then possible to have the result of the first ternary branch be cast to a `const&` without inducing UB to get the desired non-copying behavior? It seems that with the copying this map is rather not great to have.

Reply:
I recall it being during one of our review sessions, so I assume you were there.  I don't think we did a full analysis of the code to figure out exactly what went wrong, we (@gsb76, in particular,  I think) just tried it and it didn't work.

Reply:
Ah okay, then I was likely there. I'm curious about what exactly is happening so I'll try it in Wedge3D and see if I can provide some insight. It would be good to understand what the best approach here is.

Reply:
Alright, I tried this with a simple example and here are the results:

```cpp
// Gives const int&
TD<decltype(true ? static_cast<const int&>(1) : static_cast<const int&>(2))>
    aeu;

// Gives int
TD<decltype(true ? static_cast<const int&>(1) : 2)>
    aeu1;

// Gives int
TD<decltype(true ? 1 : static_cast<const int&>(2))>
    aeu2;
```

This would mean that we need to cast the result of `tan` to a `const &` and then we won't get a copy in the false branch.

Reply:
Testing indicates that that leads to a dangling reference:
```
$ g++ -std=c++14 foo.cpp
$ ./a.out 
No cast, not using temporary: allocation, memory OK
No cast, using temporary: allocation, memory OK
With cast, not using temporary: no allocation, memory OK
With cast, using temporary: allocation, dangling
$ cat foo.cpp
#include <iostream>

struct X {
  bool freed{false};
  ~X() { freed = true; }
};

void test_no_cast(bool temp) {
  const X x{};
  const auto& x2 = temp ? X{} : x;
  std::cout << "No cast, " << (temp ? "" : "not ") << "using temporary: "
            << (&x == &x2 ? "no " : "") << "allocation, "
            << (x2.freed ? "dangling": "memory OK") << "\n";
}

void test_cast(bool temp) {
  const X x{};
  const auto& x2 = temp ? static_cast<const X&>(X{}) : x;
  std::cout << "With cast, " << (temp ? "" : "not ") << "using temporary: "
            << (&x == &x2 ? "no " : "") << "allocation, "
            << (x2.freed ? "dangling": "memory OK") << "\n";
}

int main() {
  test_no_cast(false);
  test_no_cast(true);
  test_cast(false);
  test_cast(true);
}
```

Reply:
Then the cleanest is probably putting lines 37 onward into a lambda that takes the `cap_`s as const references, with an `if` statement depending on the bool. Thoughts on that idea?

Reply:
Yeah, that sounds like it would work.  The only downside is the logic gets a bit out of order, but that seems to just be how modern C++ reads and we're presumably used to it by now.

Reply:
Yea, agreed, though it's definitely something we do a lot, and given that otherwise we're copying a bunch of DataVectors, I think the trade-off is worth it.